libakumuli/storage_engine/blockstore.cpp

```diff
@@ -377,10 +324,169 @@ static u32 crc32c(const u8* data, size_t size) {
+                                             std::vector<std::string> volpaths)
+  : FileStorage::FileStorage(metapath, volpaths)
+  , db_name_(db_name)
+{
+  // nothing specific needed except calling the parent constructor
+}
+
+std::shared_ptr<ExpandableFileStorage> ExpandableFileStorage::open(std::string db_name,
+                                                                   std::string metapath,
+                                                                   std::vector<std::string> volpaths) {
+    if (volpaths.empty() || metapath.empty()) {
+      AKU_PANIC("Database file(s) doesn't exists!");
+    }
+    auto bs = new ExpandableFileStorage(db_name, metapath, volpaths);
+    return std::shared_ptr<ExpandableFileStorage>(bs);
+}
+
+bool ExpandableFileStorage::exists(LogicAddr addr) const {
+    std::lock_guard<std::mutex> guard(lock_); AKU_UNUSED(guard);
+    auto gen = extract_gen(addr);
+    auto vol = extract_vol(addr);
+    aku_Status status;
+    u32 actual_gen;
+    std::tie(status, actual_gen) = meta_->get_generation(gen);
+    if (status != AKU_SUCCESS) {
+      return false;
+    }
+    u32 nblocks;
+    std::tie(status, nblocks) = meta_->get_nblocks(gen);
+    if (status != AKU_SUCCESS) {
+      return false;
+    }
+    return actual_gen == gen && vol < nblocks;
+}
+
+std::tuple<aku_Status, std::shared_ptr<Block>> ExpandableFileStorage::read_block(LogicAddr addr) {
+    std::lock_guard<std::mutex> guard(lock_); AKU_UNUSED(guard);
+    aku_Status status;
+    auto gen = extract_gen(addr);
+    auto vol = extract_vol(addr);
+    u32 actual_gen;
+    u32 nblocks;
+    std::tie(status, actual_gen) = meta_->get_generation(gen);
+    if (status != AKU_SUCCESS) {
+      return std::make_tuple(AKU_EBAD_ARG, std::unique_ptr<Block>());
+    }
+    std::tie(status, nblocks) = meta_->get_nblocks(gen);
+    if (status != AKU_SUCCESS) {
+        return std::make_tuple(AKU_EBAD_ARG, std::unique_ptr<Block>());
+    }
+    if (actual_gen != gen || vol >= nblocks) {
+      return std::make_tuple(AKU_EUNAVAILABLE, std::unique_ptr<Block>());
+    }
+    std::vector<u8> dest(AKU_BLOCK_SIZE, 0);
+    status = volumes_[gen]->read_block(vol, dest.data());
+    if (status != AKU_SUCCESS) {
+        return std::make_tuple(status, std::unique_ptr<Block>());
+    }
+    auto block = std::make_shared<Block>(addr, std::move(dest));
+    return std::make_tuple(status, std::move(block));
+}
+
+std::unique_ptr<Volume> ExpandableFileStorage::create_new_volume(u32 id) {
+    u32 prev_id = current_volume_ - 1;
+    boost::filesystem::path prev_path(volumes_[prev_id]->get_path());
+    auto pp = prev_path.parent_path();
+    std::string basename = std::string(db_name_) + "_" + std::to_string(id) + ".vol";
+    boost::filesystem::path new_path = pp / basename;
+    Volume::create_new(new_path.c_str(), volumes_[prev_id]->get_size());
+    return Volume::open_existing(new_path.c_str(), 0);
+}
+
+void ExpandableFileStorage::update_current_volume() {
```

Comment:
I don't think that you've introduced any memory leaks.
And it looks like there is no need to lock the mutex inside this method. This method is only called from other method that already locks the mutex. When the `update_current_volume` gets called it's always holds the lock so just ignore my previous comment.